### PR TITLE
Be more specific about 'not authorized' error message #4128

### LINKED ISSUE #4128: Be more specific about 'not authorized' error message
For example, when a student tries to access a not-open session using a link, we can inform the student that the session is not open, instead of saying not authorized.

### PR DESCRIPTION
**Pull request by LowWeiLin**
Fixes #4128
Added a parameter to display the message associated with UnauthorizedAccessException to the user.

image

### CODE REVIEW

**Comment by damithc** [comment]
Can we also show the opening time and closing time in the error message?
>
>

-- 
Sent from a mobile device.

**Comment by LowWeiLin** [comment]
Do you mean the visible time? It is separate from opening time.

**Comment by damithc** [comment]
Actually it depends on what the user is trying to do. If the user is trying to submit, we should show the open and close times. If the user is trying to view results, we should inform that the results are not yet visible.

**Comment by LowWeiLin** [comment]
Hmm the error i did it for was when the session itself is not visible yet. I'll check which other errors leads to the UnauthorizedAccessException.

**Comment by damithc** [comment]
The case I'm facing: an instructor closed a session before the original closing date. Students still have the original email stating the original closing date, and they encounter the unauthorized error when they try to submit.

**Comment by LowWeiLin** [comment]
How did the instructor close the session? Moving the closing date forward still allows the student to see the feedback session although they cannot submit(they won't see the error page). Was the session deleted?
The other reason that UnauthorizedAccessException might be thrown is if the session was made private.

**Comment by LowWeiLin** [comment]
Or is it that the students encounters the error after clicking the submit button? This could happen if the page was loaded before the closing time.

**Comment by damithc** [comment]
Good point. Turns out the instructor changed both start and end dates to be in the future. So in those cases we can simply give the start date in the error message?

**Comment by LowWeiLin** [comment]
The error is due to the session not being visible(to students). Usually it is set to the opening time(when students can submit) but could be different. I'll display that date instead, since it is the one that causes the error.

**Comment by damithc** [comment]
Yes, it is set to become visible at opening time.

**Comment by LowWeiLin** [comment]
Updated
image

**Comment by damithc** [comment]
Let's rephrase it to make it sound softer. The current one sounds like the user is being blamed for doing something wrong. Also, student users may not be familiar with the concept of 'visibility'
"Sorry, this session is currently not open for submission yet. It will be open from ..."

Also, the 'logout and return' is not applicable to those who access the the page using the link?

**Comment by damithc** [comment]
This might require making the error page more configurable (and possible convert to JSTL as well?). Not sure if we do it here or as a separate PR.

**Comment by LowWeiLin** [comment]
Seems like UnauthorizedAccessException may not be the right exception to throw for this case? I can create a new exception which just shows a custom message, without the "You are not authorized to view this page."  or logout link.

**Comment by damithc** [comment]
Yes, may be as a subclass? Because at a higher level it is a case of
unauthorized access. Alternatively, you can enhance the exception to
contain more information about the cause so that we can take appropriate
action based on that information.

**Comment by LowWeiLin** [comment]
Updated

**Comment by LowWeiLin** [comment]
image

**Comment by kanghj** [comment]
Take note of the new naming convention for branches

**Comment by damithc** [comment]
Remove either 'currently' or 'yet'? Sounds odd to have both.
